_posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md

```diff
@@ -0,0 +1,89 @@
+---
+title: "Fimd out what Callbacks are defined on an Active Record model in the console"
+description: "Sometimes it's hars to know"
+layout: article
+category: ruby
+image:
+  base: "2022/find-list-debug-active-record-callbacks-in-the-console"
+  alt: "Three old payphones"
+  credit: "Pavan Trikutam"
+  source: "https://unsplash.com/photos/71CjSSB83Wo"
+---
+
+I was recently asked by my [First #RubyFriend mentee](https://firstrubyfriend.org) how to list the callbacks present on an Active Record model. I didn't know.
+
+The situation was a large legacy codebase with some large models constructed across multiple files, using various gems, custom callbacks and concerns. It was very hard to parse the callbacks by reading across multiple files.
+
+So I found a mention of debugging callbacks in the [Rails API documentation](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html#module-ActiveRecord::Callbacks-label-Debugging+callbacks) and after a little digging built a simple loop you can use to find the callbacks defined on an Active Record model.
+
+## Use...
+
+...a loop in the Rails console.
+
+Open the console.
+
+```shell
+bin/rails console
+```
+
+Paste this code into the console to see the user-defined callbacks on a model.
+
+```ruby
+YourModel.__callbacks.each_with_object(Hash.new([])) do |(k, callbacks), result|
+  next if k == :validate # ignore validations
+  
+  callbacks.each do |c|
+    # remove autosaving callbacks from result
+    next if c.filter.to_s.include?("autosave")
+    next if c.filter.to_s.include?("_ensure_no_duplicate_errors")
+
+    result["#{c.kind}_#{c.name}"] += [c.filter]
+  end
+end
+```
+
+For the `User` object (which is powered by `devise`) in my application the results look like:
```

Comment:
```suggestion
For the `User` object in my application (which is powered by [Devise](https://github.com/heartcombo/devise)) the results look like:
```

Does that look like your application is powered by Devise though? The parentheses just read a little awkwardly where they currently are...but maybe you just leave it.